src/lib.rs

```diff
@@ -0,0 +1,337 @@
+            ParseOutput::InlineParser(x) => ParseOutputType::Parser(x.into_enum()),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct InitParser;
+
+impl InitParser {
+    fn parse(self, buffer: &mut Buffer) -> ParseOutput {
+        let result = match buffer.iter().next() {
+            None => self.into_enum(),
+            Some(input) => match input {
+                b'+' => SimpleStringParser::default().into_enum(),
+                b'-' => ErrorParser::default().into_enum(),
+                b':' => IntegerParser::default().into_enum(),
+                b'$' => BulkSizeParser::default().into_enum(),
+                b'*' => ArraySizeParser::default().into_enum(),
+                _ => InlineParser::default().into_enum(),
+            },
+        };
+        let _ = buffer.pop_scanned_buffer();
+        result
+    }
+}
+
+#[derive(Debug, Default)]
+struct SimpleStringParser;
+
+impl SimpleStringParser {
+    fn parse(self, buffer: &mut Buffer) -> ParseOutput {
+        if buffer.iter().position(|x| x == b'\n') == None {
+            return self.into_enum();
+        }
+
+        let mut data = buffer.pop_scanned_buffer();
+        if data.pop() != Some(b'\n') {
+            return Error::ContentWrong.into_enum();
+        }
+        if data.pop() != Some(b'\r') {
+            return Error::ContentWrong.into_enum();
+        }
+        Message::SimpleString(data).into_enum()
+    }
+}
+
+#[derive(Debug, Default)]
+struct ErrorParser;
+
+impl ErrorParser {
+    fn parse(self, buffer: &mut Buffer) -> ParseOutput {
+        if buffer.iter().position(|x| x == b'\n') == None {
+            return self.into_enum();
+        }
+
+        let mut data = buffer.pop_scanned_buffer();
+        if data.pop() != Some(b'\n') {
+            return Error::ContentWrong.into_enum();
+        }
+        if data.pop() != Some(b'\r') {
+            return Error::ContentWrong.into_enum();
+        }
+        Message::Error(data).into_enum()
+    }
+}
+
+#[derive(Debug, Default)]
+struct IntegerParser(i64);
+
+impl IntegerParser {
+    fn parse(mut self, buffer: &mut Buffer) -> ParseOutput {
+        for i in buffer.iter() {
+            match i {
+                x @ b'0'..=b'9' => self.0 = self.0 * 10 + i64::from(x - b'0'),
+                b'-' => self.0 = -self.0,
+                b'\r' => {}
+                b'\n' => {
+                    buffer.pop_scanned_buffer();
+                    return Message::Integer(self.0).into_enum();
+                }
+                _ => return Error::ContentWrong.into_enum(),
+            }
+        }
+        self.into_enum()
+    }
+}
+
+#[derive(Debug, Default)]
+struct BulkSizeParser(isize);
+
+impl BulkSizeParser {
+    fn parse(mut self, buffer: &mut Buffer) -> ParseOutput {
+        for i in buffer.iter() {
+            match i {
+                x @ b'0'..=b'9' => self.0 = self.0 * 10 + isize::from(x - b'0'),
+                b'-' => self.0 = -self.0,
```

Comment:
bulk string's size gets only one way to negative: the whole bulk is `$-1\r\n`. So you should make it faster or test it.